pkg/source/source.go

```diff
@@ -14,7 +14,7 @@ type Source string
 
 const (
 	// Unspec is used when the source is unspecified
-	Unspec Source = "unspec"
```

Comment:
The remaining question to be resolved is how to fix `TestNodeLifecycle`.

This discussion was started in https://github.com/cilium/cilium/pull/35810#discussion_r1836161757 and then continued in https://github.com/cilium/cilium/pull/35810#discussion_r1850887796.

Recap: `TestNodeLifecycle` calls `func (m *manager) NodeUpdated(n nodeTypes.Node)`, which in turn calls `AllowOverwrite` with the `existing` source being `Unspec` (because it is the default value returned by `func (s prefixInfo) Source() source.Source`) and the `new` source being `""`. Passing an empty string as source does not seem to be a proper usage and there are two ways to fix this:

1. Redefine `source.Unspec` to equal `""`, which unifies these two undefined values.
2. Use an explicit source instead of the default empty string in `TestNodeLifecycle`.

Assuming that two undefined values are not needed (one definitely is needed as per https://github.com/cilium/cilium/blob/b424bfda4f686fb34271c91d292372f4353477b2/pkg/ipcache/types.go#L39-L42), I implemented solution (1) above because with (2) we can make the test green but break the actual functionality.

Indeed, following up on the earlier investigation from https://github.com/cilium/cilium/pull/35810#discussion_r1851721413, I checked that `StartDiscovery` calls `NodeUpdated` with `localNode.Node`, which originates from `n.localNodeStore`, and `localNodeStore` is set in `NewNodeDiscovery` as `lns *node.LocalNodeStore`, which is [provided](https://github.com/cilium/cilium/blob/4b2264b84b537ef329eaa0d77c0a1ffb5b3923e1/pkg/node/local_node_store.go#L53) via `Hive` from `func NewLocalNodeStore(params LocalNodeStoreParams) (*LocalNodeStore, error)`, which does **NOT** set the `Source` field of the `Node` field of the `value LocalNode` field. So in production the source is an empty string just like in the test.

Thinking about it now, I wonder that perhaps the best solution would be the following:

 3. Use an explicit source in both `TestNodeLifecycle` and `NewLocalNodeStore`.

Could you please let me know what you think?

Reply:
I think it would be reasonable to explicitly set the `LocalNode`'s source as `Local`. That's what it's set to in `daemon/cmd/local_node_sync.go` anyway.

Reply:
Oh, I see it now, thanks Jussi!

`func NewLocalNodeStore(params LocalNodeStoreParams) (*LocalNodeStore, error)` creates `LocalNodeStore` whose `value LocalNode` has `""` as `source` (because it's not explicitly specified), but then immediately (unless `params.Sync` is nil) it adds an `OnStart` hook that calls `InitLocalNode()` and starts a new go-routine running `SyncLocalNode()` (both functions from `daemon/cmd/local_node_sync.go` mentioned in the previous comment). Among these two, `InitLocalNode()` [sets the source as](https://github.com/cilium/cilium/blob/60f639c555a8ffb56ba1830b35bd1737c7030e28/daemon/cmd/local_node_sync.go#L50) `source.Local` and `SyncLocalNode()` [sets it as](https://github.com/cilium/cilium/blob/60f639c555a8ffb56ba1830b35bd1737c7030e28/daemon/cmd/local_node_sync.go#L272) `source.Kubernetes`. So it should be safe to initialize the `source` field of `value LocalNode` as anything with precedence lower than or equal to these of `source.Local` and `source.Kubernetes` (`source.Local` has higher precedence than `source.Kubernetes`, so I'd rather avoid it). Since it used to be `""`, I think I'll just go with `source.Unspec`.

Reply:
Could you please also have a look and let me know what you think, @joestringer?